src/main/java/org/jenkinsci/plugins/azurekeyvaultplugin/AzureCredentialsProvider.java

```diff
@@ -0,0 +1,105 @@
+import com.microsoft.azure.keyvault.models.CertificateItem;
+import com.microsoft.azure.keyvault.models.SecretBundle;
+import com.microsoft.azure.keyvault.models.SecretItem;
+import com.microsoft.jenkins.keyvault.SecretCertificateCredentials;
+import com.microsoft.jenkins.keyvault.SecretStringCredentials;
+import edu.umd.cs.findbugs.annotations.NonNull;
+import edu.umd.cs.findbugs.annotations.Nullable;
+import hudson.Extension;
+import hudson.model.ItemGroup;
+import hudson.model.ModelObject;
+import hudson.security.ACL;
+import hudson.util.Secret;
+import java.time.Duration;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.List;
+import java.util.concurrent.TimeUnit;
+import java.util.function.Supplier;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+import jenkins.model.GlobalConfiguration;
+import jenkins.model.Jenkins;
+import org.acegisecurity.Authentication;
+
+
+@Extension
+public class AzureCredentialsProvider extends CredentialsProvider {
+    private static final Logger LOG = Logger.getLogger(AzureCredentialsProvider.class.getName());
+
+    private final AzureCredentialsStore store = new AzureCredentialsStore(this);
+
+    private final Supplier<Collection<IdCredentials>> credentialsSupplier =
+            memoizeWithExpiration(AzureCredentialsProvider::fetchCredentials, Duration.ofMinutes(5));
+
+    private static <T> Supplier<T> memoizeWithExpiration(Supplier<T> base, Duration duration) {
+        return Suppliers.memoizeWithExpiration(base::get, duration.toMillis(), TimeUnit.MILLISECONDS)::get;
+    }
+
+    @NonNull
+    @Override
+    public <C extends Credentials> List<C> getCredentials(@NonNull Class<C> aClass, @Nullable ItemGroup itemGroup,
+                                                          @Nullable Authentication authentication) {
+        if (ACL.SYSTEM.equals(authentication)) {
+            final ArrayList<C> list = new ArrayList<>();
+            for (IdCredentials credential : credentialsSupplier.get()) {
+                if (aClass.isAssignableFrom(credential.getClass())) {
+                    // cast to keep generics happy even though we are assignable..
+                    list.add(aClass.cast(credential));
+                }
+                LOG.log(Level.FINEST, "getCredentials {0} does not match", credential.getId());
+            }
+            return list;
+        }
+
+        return Collections.emptyList();
+    }
+
+    private static Collection<IdCredentials> fetchCredentials() {
+        AzureKeyVaultGlobalConfiguration azureKeyVaultGlobalConfiguration = GlobalConfiguration.all().get(AzureKeyVaultGlobalConfiguration.class);
+        if (azureKeyVaultGlobalConfiguration == null) {
+            throw new AzureKeyVaultException("No global key vault url configured.");
+        }
+        KeyVaultClient client = new KeyVaultClient(new AzureKeyVaultImdsCredential());
```

Comment:
One problem here is that it only supports MSI now. I have not succeeded to get credentials with id out of a build.

Reply:
is it something that could be configured globally? not per build?

Reply:
I can get the `credentialID` from the global configuration. But I cannot get the credential like code https://github.com/jenkinsci/azure-keyvault-plugin/blob/f2f54bd82f9fb2f9ac48f718bf00802d40cae502/src/main/java/org/jenkinsci/plugins/azurekeyvaultplugin/AzureKeyVaultCredentialRetriever.java#L20-L59 without `Run` object.

Reply:
you should be able to pass null or add another override that doesn't require a run